Misc/NEWS.d/next/Core and Builtins/2019-02-12-20-16-34.bpo-35961.7f7Sne.rst

```diff
@@ -0,0 +1,2 @@
+Fix a crash in slice_richcompare(): use strong references rather than borrowed
```

Comment:
The original code was using "stolen" references.

Reply:
For me, it's a borrowed reference, not a stolen reference: objects are still referenced by v and w, no?

Reply:
Yes, but `v` and `w` no longer _own_ the references.  Ownership has been stolen by the tuple.  [The docs](https://docs.python.org/3/c-api/intro.html#reference-count-details) leave no doubt about this: 

>Few functions steal references; the two notable exceptions are PyList_SetItem() and PyTuple_SetItem(), which steal a reference to the item (but not to the tuple or list into which the item is put!). These functions were designed to steal a reference ...

Reply:
Well, technically v and w still own the reference, that's why the function needs to clear the reference after PyObject_RichCompare() call:

```
    PyTuple_SET_ITEM(t1, 0, NULL);
    PyTuple_SET_ITEM(t1, 1, NULL);
    PyTuple_SET_ITEM(t1, 2, NULL);
    PyTuple_SET_ITEM(t2, 0, NULL);
    PyTuple_SET_ITEM(t2, 1, NULL);
    PyTuple_SET_ITEM(t2, 2, NULL);
```

I'm not sure that the subtle difference really matters here :-)

Reply:
Why argue about this?  The docs couldn't possibly be more explicit about that `PyTuple_SetItem` steals a reference, and two people (including me) so far have pointed out that "stolen" is the correct terminology here.

The code you just posted was Guido's attempt to _return_ ownership of the references to the slice objects.  He knew darned well that he stole them to begin with.  What he overlooked is that gc can blow up if there are more references visible to gc than an object's refcount claims exist.

The code setting the slots to `NULL` is needed precisely _because_ the tuple stole the references:  without setting to `NULL`, decref'ing the tuple would in turn reduce the refcounts of objects in the tuple slots.  That's essentially what "ownership" means:  the tuple is responsible for decref'ing its items when the tuple goes away.  He had to break the tuple's ownership of its items because the tuple's references were stolen to begin with.

Of course it's not a big deal if a check-in comment is misleading about a technical detail, but it's not doing anyone any favors either :wink:.

Reply:
API who steal references must die. https://pythoncapi.readthedocs.io/bad_api.html#borrowed-references

Reply:
I modified the NEWS entry to replace borrowed with stolen ;-)